full-service/src/db/encryption_provider.rs

```diff
@@ -0,0 +1,304 @@
+                    *cur_password_hash = password_hash.to_vec();
+                } else {
+                    return Err(WalletDbError::PasswordFailed);
+                }
+            }
+            EncryptionState::Unencrypted => {
+                log::info!(
+                    self.logger,
+                    "DB is in a bad state where some accounts are unencrypted. Please run change_password or contact support to run migration tool."
+                );
+                return Err(WalletDbError::BadEncryptionState);
+            }
+        }
+        Ok(())
+    }
+
+    pub fn change_password(
+        &self,
+        old_password_hash: &[u8],
+        new_password_hash: &[u8],
+        conn: &PooledConnection<ConnectionManager<SqliteConnection>>,
+    ) -> Result<(), WalletDbError> {
+        // Check whether encrypted, and if so, then verify old password. For any other state, set_password to new.
+        match EncryptionIndicator::get_encryption_state(&conn)? {
+            EncryptionState::Empty => {
+                log::info!(
+                    self.logger,
+                    "Database has never been locked. Setting password with new password."
+                );
+                self.set_password_hash(new_password_hash, &conn)?;
+            }
+            EncryptionState::Encrypted => {
+                log::debug!(self.logger, "Database is locked. Verifying old password.");
+                // Attempt to decrypt the test value to confirm if password is correct
+                let expected_val =
+                    Self::encrypt_with_password(ENCRYPTION_VERIFICATION_VAL, old_password_hash)?;
+                if EncryptionIndicator::verify_encrypted_value(&expected_val, &conn)? {
+                    // Set password to new password
+                    self.set_password_hash(new_password_hash, &conn)?;
+                } else {
+                    return Err(WalletDbError::PasswordFailed);
+                }
+            }
+            EncryptionState::Unencrypted => {
+                log::info!(
+                    self.logger,
+                    "Database is unencrypted. Setting password with new password."
+                );
+                self.set_password_hash(new_password_hash, &conn)?;
+            }
+        }
+        Ok(())
+    }
+
+    pub fn encrypt(&self, plaintext_bytes: &[u8]) -> Result<Vec<u8>, WalletDbError> {
+        Ok(Self::encrypt_with_password(
+            plaintext_bytes,
+            &self.get_password_hash()?,
+        )?)
+    }
+
+    pub fn decrypt(&self, ciphertext: &[u8]) -> Result<Vec<u8>, WalletDbError> {
+        Ok(Self::decrypt_with_password(
+            ciphertext,
+            &self.get_password_hash()?,
+        )?)
+    }
+
+    /// Encrypt data.
+    pub fn encrypt_with_password(
+        plaintext_bytes: &[u8],
+        password_hash: &[u8],
+    ) -> Result<Vec<u8>, WalletDbError> {
+        let (key, nonce) = Self::expand_password(password_hash)?;
```

Comment:
This looks like it uses the same key and nonce for every encryption done for a given password, and I believe AES-GCM breaks if a (key,nonce) pair is ever reused, e.g. https://chris-wood.github.io/2016/12/25/GCM-SIV.html .

If so, we could choose a more nonce-misuse-resistant algorithm (AES-GCM-SIV, ChaCha20?), or be careful to use a distinct nonce for each encryption. I think its common to use a counter as the nonce (incremented for each encryption, so always unique), although that will take more book-keeping on the DB side to maintain a global counter for every encryption.